### PR TITLE
Periodic correction for the constraintsIB method

### PR DESCRIPTION
**Pull request by hongk45**
corrects the calculations of the COM, MOI and related quantities in the presence of periodic boundary for constraintIB

### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [ ] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [ ] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [ ] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [ ] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

### CODE REVIEW

**Comment by hongk45** [comment]
@drwells: just decided to create a brand-new PR to make all look cleaner

**Comment by hongk45** [comment]
@drwells : how do I make a new entry to IBAMR/doc/news/changes/? create a new PR?

**Comment by drwells** [comment]
Just add a new file into `doc/news/changes/minor` to this PR (it should look like https://github.com/IBAMR/IBAMR/blob/master/doc/news/changes/minor/20201016DavidWells, which is also a bug fix). Make sure the file is named in the same way (YYYYMMDDName).

File: src/IB/ConstraintIBMethod.cpp
Comment (drwells) [inline src/IB/ConstraintIBMethod.cpp:221]: These need to be after `d_center_of_mass_new` to match the declaration order in the header.

**Review by drwells — CHANGES_REQUESTED** [review]
Looks good but we need to fix a compiler warning.

**Comment by drwells** [comment]
All the tests pass so this is probably fine. @amneetb - could you take a second look since you understand this class best?

**Review by drwells — APPROVED** [review]

**Comment by boyceg** [comment]
@drwells, is it worth squashing some commits before merging?

**Comment by drwells** [comment]
Cleaning up history is nice but ultimately optional. If @hongk45 knows how to squash commits that would be great - otherwise we can go with it as-is.

**Comment by hongk45** [comment]
@drwells, @boyceg: I tried a bit to clean up the history. Hope it appears cleaner now.

**Comment by drwells** [comment]
@hongk45 I'm not sure things are the way you intended for them to be - we went from six commits to thirteen.

The current state of the history is now problematic since it contains a merge commit. In particular, if we were to merge this PR as it stands, the commits here would be intercalated with those on `master`, which would make `git bisect` no longer work. More exactly: I'm fine with not squashing commits but we cannot have 'merge into master' commits like this in PRs since it makes using other git tools impossible.

The best way to fix this situation is to completely rewrite history and force push. More exactly:
1. Do a soft reset so that the commits are gone but the changes are still present.
2. Add the changes again and write a single commit containing all of them.

the following git commands should resolve this problem:
```
git checkout periodic-correction-constraintIB
git merge master
git reset --soft master
git commit -m "Periodic correction for the ConstraintIB method"
```
In order:
1. Move to the branch
2. If you have updated master then we merge those changes in, in order. This gets around the problem with intercalated changes to `master` making `git rebase` not work.
3. Keep the changes but reset `HEAD` to master (i.e., remove all commits but not the changes)
4. Recommit all the changes into a single commit

Do this and then force push and we should be set :+1:

**Review by drwells — CHANGES_REQUESTED** [review]
No merge commits, please.

**Comment by amneetb** [comment]
Do we have an example to test the new functionality?

**Comment by hongk45** [comment]
@drwells: followed your steps, but the problem seems to be persistent. Here is what I did:
---------------
git checkout periodic-correction-constraintIB
git merge master
git reset --soft master
git commit -m "Periodic correction for the ConstraintIB method"
git add doc/news/changes/minor/20201201HongNguyen
git add src/IB/ConstraintIBMethod.cpp
git add src/IB/ConstraintIBMethod.cpp
git push origin periodic-correction-constraintIB
--------------------
Am I missing anything?

**Comment by boyceg** [comment]
> Do we have an example to test the new functionality?

This is a good point. If it has not already been done, we should verify that there are not any artifacts when structures "pass through" the periodic boundary. I think that ConstraintIB will be less susceptible to this than formulations designed for flexible structures, but it is valuable to check.

One thing that could be done is to compare lift/drag forces for a body that is dragged through the domain. In a fully periodic domain, the traces should be identical if the structure is offset by a distance that is an integer number of Cartesian grid cells in any direction.

**Comment by boyceg** [comment]
> ## @drwells: followed your steps, but the problem seems to be persistent. Here is what I did:
> ## git checkout periodic-correction-constraintIB
> git merge master
> git reset --soft master
> git commit -m "Periodic correction for the ConstraintIB method"
> git add doc/news/changes/minor/20201201HongNguyen
> git add src/IB/ConstraintIBMethod.cpp
> git add src/IB/ConstraintIBMethod.cpp
> git push origin periodic-correction-constraintIB
> Am I missing anything?

@hongk45: Sorry for dragging you into this git mess. 😬 

@drwells: At this point, would you be able to fix up the history for @hongk45?

**Comment by amneetb** [comment]
> One thing that could be done is to compare lift/drag forces for a body that is dragged through the domain. In a fully periodic domain, the traces should be identical if the structure is offset by a distance that is an integer number of Cartesian grid cells in any direction.

This is a very good example. The simulation can be set following sec 5.1.2 of:
https://www.sciencedirect.com/science/article/pii/S0021999117305016
https://arxiv.org/pdf/1704.00239.pdf

Another example would be 2d eel swimming. One can set the eel to cross the periodic boundary during its swimming path, and check that if its velocity or displacement remains the same or not, when it does not cross the periodic boundary.

**Comment by boyceg** [comment]
> Another example would be 2d eel swimming. One can set the eel to cross the periodic boundary during its swimming path, and check that if its velocity or displacement remains the same or not, when it does not cross the periodic boundary.

I think that you would want to do the same basic check, however --- have the initial position vary by some integer number of grid cells and verify that the kinematics are identical.

**Comment by boyceg** [comment]
One quick comment on "identical" results.

In principle, when using periodic boundaries and a uniform (not adaptive!) Cartesian grid, you should be able to offset the structure by an integer number of Cartesian grid cells and get identical results.

It is important to use a uniform Cartesian grid because the AMR grid generator may not generate identical (up to a translation) locally refined regions for the offset positions of the structure, and it _will not_ generate identical locally refined grids as the structure approaches / passes through the periodic boundary.

Further, even on a uniform Cartesian grid, there are some ordering-dependent parts of the solvers that could generate slightly different results for the offset configurations. These differences should be small compared to the overall forces and/or trajectories.

**Comment by drwells** [comment]
@hongk45 Did you force push when you updated the branch on GitHub? The last command shouldn't work since you rewrote the git history in a conflicting way. Let's hold off on fixing this until we determine what sort of example or test we should add.

**Comment by boyceg** [comment]
@hongk45, were you able to add a test to verify correctness? Note that the test does not need to be particularly well resolved --- we just want something to confirm that the method does not "see" the periodic boundary.

**Comment by hongk45** [comment]
@boyceg: Yeah, I actually simulated free swimming of a  single deforming structure in the regime of not too high Re where there is no vortex shedding. When I chose the physical box sufficiently large, then nothing peculiar appeared when the structure crossed the periodic box, for example the swimming velocity. 

I prescribed both shape and (deforming) velocity and found the same swimming velocity with "CONSTRAINT_VELOCITY" or "CONSTRAINT_POSITION". However, if the calculation of COM was not fixed and "CONSTRAINT_POSITION" was applied, when the structure was getting close to the boundary and wanted to pass, it could not do that because a wrong center of mass was used to update its position. 

In this example, I only have translational degree of freedom. If the structure is allowed to rotate, then, of course, we need a correct value of MOI.

**Comment by drwells** [comment]
This was on pause over the holiday but I'd like to get it in ASAP.

@hongk45 by 'test' we mean an integration test that doesn't necessarily run with physical parameters but still demonstrates that some part of the code should work - in particular it should run in less than a minute and use the smallest grid it can to get the point across. Can you modify your code to do this? If so, are you willing to add it to IBAMR to make sure that ConstraintsIB continues to work with periodic BCs?

**Comment by hongk45** [comment]
> 
> 
> This was on pause over the holiday but I'd like to get it in ASAP.
> 
> @hongk45 by 'test' we mean an integration test that doesn't necessarily run with physical parameters but still demonstrates that some part of the code should work - in particular it should run in less than a minute and use the smallest grid it can to get the point across. Can you modify your code to do this? 
>
@drwells: sorry for my slow reply. Actually, I don't mind modifying my example to serve as a test for this additional functionality.  
>If so, are you willing to add it to IBAMR to make sure that ConstraintsIB continues to work with periodic BCs?
>
You mean it is possible for my example to be added to the master branch if it works well, or something else. Sorry I am a bit confused here.

Anyway, please help me to understand why the current example codes in examples/ConstraintIB/ can not be used to test this. I think 'impulsively_started_cylinder', 'falling_sphere' or 'oscillating_rigid_cylinder' might be a good tester with a little change of input parameters.

**Comment by drwells** [comment]
We would add the test to this PR so that we merge both the bug fix and the test that verifies the bug fix at the same time.

> Anyway, please help me to understand why the current example codes in examples/ConstraintIB/ can not be used to test this. I think 'impulsively_started_cylinder', 'falling_sphere' or 'oscillating_rigid_cylinder' might be a good tester with a little change of input parameters.

If all we need to verify that the bug is fixed is a changed input file then that's even better. Adding a new test or example requires modifying the build system - I can take care of that. Could you attach the input file to a GitHub comment in this PR?

**Comment by drwells** [comment]
@hongk45 Lets try to wrap this up soon. Do you still have the input file you used to show that these changes work? If so, would you be willing to send it to me (either by email or by posting it here)?

**Comment by hongk45** [comment]
> 
> 
> @hongk45 Lets try to wrap this up soon. Do you still have the input file you used to show that these changes work? If so, would you be willing to send it to me (either by email or by posting it here)?

hi @drwells, i am currently fully occupied by something else. I will get back to this by the end of this week or next.

**Comment by drwells** [comment]
Not a problem - let me know when you're ready to wrap this up.

**Comment by hongk45** [comment]
> 
> 
> Not a problem - let me know when you're ready to wrap this up.

hi @drwells, just want to let you know that i will get back to this ASAP

**Comment by hongk45** [comment]
Hi @drwells, I have here 2 input files that might be useful to verify that this fix actually works.
These are for oscillating_rigid_cylinder example:
1) [input2d_oscillating_rigid_cylinder_nonPBC.txt](https://github.com/IBAMR/IBAMR/files/5974146/input2d_oscillating_rigid_cylinder_nonPBC.txt)
This file only applies PBC along y, but not along x. The cylindrer is placed at the center of the box which is large enough to be considered as infinite.

2) [input2d_oscillating_rigid_cylinder_PBC.txt](https://github.com/IBAMR/IBAMR/files/5974148/input2d_oscillating_rigid_cylinder_PBC.txt)
This file applies PBC along both x and y. The cylinder is placed right at the boudary at the beginning. The point here is to make the structure to continually cross the periodic boundary. The box size is the same as in previous case. 


We would expect that the 2 setups will produce identical results, e.g the drag force. And indeed, this is the case.
You can also check that this fix will produce the correct result for the cylinder's COM while previous unfixed code cannot.

Let me know if you need more example.

**Review by drwells — APPROVED** [review]

**Comment by drwells** [comment]
All the tests pass and this looks good to me so lets merge.

**Comment by drwells** [comment]
@hongk45 Thanks again for fixing this!

**Pull request merged**
